core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt

```diff
@@ -636,4 +636,330 @@ class BlockAvailabilityTests {
+                explorer,
+                Offset(0.meters),
+                explorer.getSimulatedLength(),
+                0.0
+            ) as BlockAvailabilityInterface.Unavailable
+        assertEquals(minDelay, res.duration)
+        assertEquals(plannedStepOffset.distance, res.firstConflictOffset.distance)
+    }
+
+    /**
+     * Passing before step returns delay needed to respect its timing data. This delay makes the
+     * train pass after another step: should return Unavailable(infinity, ...).
+     */
+    @Test
+    fun testPassingBeforeStepCreatesDelayMakesAvailableStepUnavailable() {
+        val explorer = makeExplorer(5, 1)
+        val plannedStepOffset = Offset<Block>(0.meters)
+        val secondPlannedStepOffset = Offset<Block>(100.meters)
+        val timeAtZoneEnd = explorer.interpolateDepartureFromClamp(explorer.getSimulatedLength())
+        val steps =
+            listOf(
+                // We need to add delay = timeAtZoneEnd
+                STDCMStep(
+                    listOf(PathfindingEdgeLocationId(blocks[0], plannedStepOffset)),
+                    null,
+                    false,
+                    PlannedTimingData(timeAtZoneEnd.seconds, 0.seconds, 0.seconds)
+                ),
+                // If we add delay = timeAtZoneEnd, this step is not respected
+                STDCMStep(
+                    listOf(PathfindingEdgeLocationId(blocks[0], secondPlannedStepOffset)),
+                    null,
+                    false,
+                    PlannedTimingData(0.seconds, 0.seconds, timeAtZoneEnd.seconds)
+                )
+            )
+        val availability = makeBlockAvailability(infra, listOf(), listOf(), steps)
+        val res =
+            availability.getAvailability(
+                explorer,
+                Offset(0.meters),
+                explorer.getSimulatedLength(),
+                0.0
+            ) as BlockAvailabilityInterface.Unavailable
+        // The delay added to solve the first step makes the train pass after the second step =>
+        // infinity + second step offset
+        assertEquals(POSITIVE_INFINITY, res.duration)
+        assertEquals(plannedStepOffset.distance, res.firstConflictOffset.distance)
+    }
+
+    /** Unavailable step has priority over available conflicts. */
+    @Test
+    fun testPassingBeforePlannedStepWithoutUnavailableConflicts() {
+        val explorer = makeExplorer(5, 1)
+
+        val plannedStepOffset = Offset<Block>(100.meters)
+        val timeAtStep = explorer.interpolateDepartureFromClamp(Offset(plannedStepOffset.distance))
+        val timeAtZoneEnd = explorer.interpolateDepartureFromClamp(explorer.getSimulatedLength())
+        val steps =
+            listOf(
+                STDCMStep(
+                    listOf(PathfindingEdgeLocationId(blocks[0], Offset(100.meters))),
+                    null,
+                    false,
+                    PlannedTimingData(timeAtZoneEnd.seconds, 0.seconds, 30.seconds)
```

Comment:
If I get this right the train arrives as the step is available, it doesn't match the comments

Reply:
No. TimeAtZoneEnd is the end of the time at the end of the simulated length, which is the first zone. The end of the zone is not 100.meters, it's a lot larger. Time at zone end is like 160s, time at 100.meters is 3.3333s.